Added Blue Score the respective  folders

karthikrangasai opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #352 .

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

SkafteNicki commented on torchmetrics/functional/__init__.py:
```diff
@@ -55,3 +54,4 @@
 from torchmetrics.functional.retrieval.recall import retrieval_recall  # noqa: F401
 from torchmetrics.functional.retrieval.reciprocal_rank import retrieval_reciprocal_rank  # noqa: F401
 from torchmetrics.functional.self_supervised import embedding_similarity  # noqa: F401
+from torchmetrics.functional.text import bleu_score  # noqa: F401
```
full path :]
```suggestion
from torchmetrics.functional.text.bleu import bleu_score  # noqa: F401
```

SkafteNicki commented on torchmetrics/functional/__init__.py:
```diff
@@ -33,7 +33,6 @@
 from torchmetrics.functional.classification.specificity import specificity  # noqa: F401
 from torchmetrics.functional.classification.stat_scores import stat_scores  # noqa: F401
 from torchmetrics.functional.image_gradients import image_gradients  # noqa: F401
-from torchmetrics.functional.nlp import bleu_score  # noqa: F401
```
@Borda should we deprecate this?
Only influences people using the full import path
`from torchmetrics.functional.nlp import bleu_score`
whereas
`from torchmetrics.functional import bleu_score`
will still work

Borda replied:
yep, rather make simple deprecation loop

SkafteNicki reviewed (commented):
Hi @karthikrangasai,
Looks good until now. Could you also:
* Add corresponding tests / move the onces in `tests/functiona/test_nlp.py` to `tests/text/test_bleu.py`
* Add /update references in the docs

karthikrangasai commented:
@SkafteNicki 
Yeah, I will do the same.

karthikrangasai commented:
I have refactored the code such that the Class implementation uses the functional implementation of the BLEU score. I have moved the tests for the functional implementation to the respective directory. 

Regarding the tests for the `BLEUScore` class, I have doubts about using the existing `MetricTester` because it accepts `preds` and `targets` as `torch.Tensor` type but the inputs for the BLEU Score metric are an iterable of the strings.

One hack I thought we could use is to provide the iterable of strings as `**mertic_args` and leave `preds` and `targets` as `None` - but this failed all the tests.

Other option I see is to implement a separate `TextMetricTester` for all Text based metrics. I am not too sure about this methods.

How do I proceed with the tests for Class based Text Metrics?

SkafteNicki commented on torchmetrics/text/bleu.py:
```diff
@@ -0,0 +1,82 @@
+        self.add_state("numerator", torch.zeros(self.n_gram), dist_reduce_fx="sum")
+        self.add_state("denominator", torch.zeros(self.n_gram), dist_reduce_fx="sum")
+
+    def update(self, reference_corpus, translate_corpus) -> None:
```
missing typing for `reference_corpus` and `translate_corpus`

SkafteNicki commented on torchmetrics/text/bleu.py:
```diff
@@ -0,0 +1,82 @@
+
+class BLEUScore(Metric):
+    """
+    Calculate BLEU score of machine translated text with one or more references.
```
```suggestion
    Calculate `BLEU score <https://en.wikipedia.org/wiki/BLEU>`_ of machine translated text with one or more references.
```

SkafteNicki commented on torchmetrics/functional/text/bleu.py:
```diff
@@ -115,3 +101,48 @@ def bleu_score(
+
+    c, r = _bleu_score_update(reference_corpus, translate_corpus, numerator, denominator, c, r, n_gram)
+
+    trans_len = c.clone().detach()
+    ref_len = r.clone().detach()
```
why is the `clone().detach()` necessary?

karthikrangasai replied:
The initial implementation had this, I'm not really sure why we are doing it. It just makes a copy that is not considered in the computation graph of the model. We can use the same `c, r` I guess. Please let me know what I should do.

karthikrangasai replied:
Not doing the `clone().detach()` operation doesn't affect the tests.

SkafteNicki replied:
then I would remove them :]

SkafteNicki commented on torchmetrics/functional/text/bleu.py:
```diff
@@ -97,9 +73,19 @@ def bleu_score(
+    ref_len: Tensor,
+    numerator: Tensor,
+    denominator: Tensor,
+    c: float,
+    r: float,
```
```suggestion
    c: Tensor,
    r: Tensor,
```

SkafteNicki commented on torchmetrics/functional/text/bleu.py:
```diff
@@ -45,39 +45,15 @@ def _count_ngram(ngram_input_list: Sequence[str], n_gram: int) -> Counter:
+    translate_corpus: Sequence[Sequence[str]],
+    numerator: Tensor,
+    denominator: Tensor,
+    c: float,
+    r: float,
```
```suggestion
    c: Tensor,
    r: Tensor,
```

SkafteNicki commented on torchmetrics/functional/text/bleu.py:
```diff
@@ -115,3 +101,48 @@ def bleu_score(
+    smooth: bool = False
+) -> Tensor:
+    """
+    Calculate BLEU score of machine translated text with one or more references
```
```suggestion
    Calculate `BLEU score <https://en.wikipedia.org/wiki/BLEU>`_ of machine translated text with one or more references
```

SkafteNicki commented on torchmetrics/text/bleu.py:
```diff
@@ -0,0 +1,82 @@
+        n_gram:
+            Gram value ranged from 1 to 4 (Default 4)
+        smooth:
+            Whether or not to apply smoothing – Lin et al. 2004
```
can you insert the reference at the button instead
```suggestion
            Whether or not to apply smoothing – see [2]
```

SkafteNicki reviewed (commented):
mix of comments

SkafteNicki commented on tests/text/test_blue.py:
```diff
@@ -55,28 +57,67 @@
         smoothing_function=smooth_func,
     )
-    pl_output = bleu_score([HYPOTHESIS1], [[REFERENCE1, REFERENCE2, REFERENCE3]], n_gram=n_gram, smooth=smooth)
+    pl_output = bleu([[REFERENCE1, REFERENCE2, REFERENCE3]], [HYPOTHESIS1])
```
would it be possible for you to expand the test to include multiple batches, like:
```python
for batch in batches:
    blue.update(*batch)
pl_output = blue.compute()
```
so it is also tested that accumulation in the modular metric works

SkafteNicki commented on torchmetrics/functional/nlp.py:
```diff
@@ -69,49 +48,28 @@ def bleu_score(
         >>> reference_corpus = [['there is a cat on the mat'.split(), 'a cat is on the mat'.split()]]
         >>> bleu_score(translate_corpus, reference_corpus)
         tensor(0.7598)
+
```
can you include a warning similar to this https://github.com/PyTorchLightning/metrics/blob/a4d2a7a81ba482f28edd30bd76abc3cffbaca548/torchmetrics/functional/regression/mean_relative_error.py#L43-L44 in the docstring

SkafteNicki reviewed (commented):
minor comments

karthikrangasai commented:
I have pushed the updated changes.

Borda commented on docs/source/references/functional.rst:
```diff
@@ -281,17 +281,6 @@ ssim [func]
 
-
-***
-NLP
```
shall we call it NLP or Text?

karthikrangasai replied:
I am not sure about this. NLP also includes speech processing. So, if we are to add those metrics as well then we can call it NLP.

Borda replied:
well, speech processing you mean conversion from audio > text and back, right?
bu then you still shall measure the quality against each independently as your prediction and target are always other audio or text, right? so we can split NLP as text and audio... :rabbit: 
cc: @SkafteNicki @maximsch2

SkafteNicki replied:
I prefer text, as the data modality that bleu works on is text, similar to how we have grouped other metrics based on data modality.
Also, this does not matter for the end users, as all modular metrics can just be imported with `from torchmetrics import *` and functional `from torchmetrics.functional import *

Borda commented on tests/text/test_blue.py:
```diff
@@ -42,6 +44,11 @@
 LIST_OF_REFERENCES = [[REF1A, REF1B, REF1C], [REF2A]]
 HYPOTHESES = [HYP1, HYP2]
```
```suggestion
LIST_OF_REFERENCES = ((REF1A, REF1B, REF1C), (REF2A))
HYPOTHESES = (HYP1, HYP2)
```
rather immutable, and so rename the `LIST_` too

karthikrangasai replied:
`(REF2A)` doesn't make it a tuple and maintains same datatype, hence I made it `tuple(REF2A)`

Borda commented on torchmetrics/functional/nlp.py:
```diff
@@ -69,49 +48,31 @@ def bleu_score(
-    else:
-        precision_scores = numerator / denominator
+    trans_len = tensor(0, dtype=torch.float)
+    ref_len = tensor(0, dtype=torch.float)
```
just call the new function here, so it is tiny wrapper

karthikrangasai commented:
@Borda the tests are failing on Github only because it says list type is not hashable. They run fine on my system.

Borda commented:
> @Borda the tests are failing on Github only because it says list type is not hashable. They run fine on my system.

and what is your system? I assume list shall not be hashable anywhere as it immutable
https://stackoverflow.com/questions/42203673/in-python-why-is-a-tuple-hashable-but-not-a-list

karthikrangasai commented:
Yeah, it was my mistake. I was running `python -m pytest -v tests` but I had to run `python -m pytest -v torchmetrics tests`.
I have fixed the failing tests now. I had forgot to update examples in the docstring and they were failing.

Borda commented:
> I have fixed the failing tests now. I had forgot to update examples in the docstring and they were failing.

cool, mind commit your fix? :rabbit:
